Incorporating Bergsten et al. (2022)

gbergsten opened the pull request:
Added a new `create_planets_bergsten` function with updated small planet occurrence rates, now set as the default for Generators. Also included legacy tools to swap back in the old `create_planets_SAG13` function. Added a new Tutorial notebook showing these new tools.

gbergsten committed:
Incorporating Bergsten et al. (2022)

New create_planets_bergsten function with updated small planet occurrence rates, now set as the default for Generators. Legacy tools to swap back in the old create_planets_SAG13 function. Example notebook showing the new tools.

matiscke commented on bioverse/generator.py:
```diff
@@ -89,6 +89,25 @@ def insert_step(self, function, idx=None, filename=None):
         idx = len(self.steps) if idx is None else idx
         self.steps.insert(idx,step)
+        
+    def replace_step(self, new_function, idx, new_filename=None):
```
what is this used for?

matiscke replied:
I think I understand now. Other functions can also be used by building generators from scratch instead of using one of the default ones. We should still keep this useful function.

matiscke commented on bioverse/functions.py:
```diff
@@ -177,6 +177,135 @@ def read_stellar_catalog(d, filename=DATA_DIR+'LUVOIR_targets.dat', d_max=30., T
+    xv, yv = np.meshgrid(P, R)
+    
+    # Generate probability grids in bins of stellar mass
+    for i,_ in enumerate(F_0):
```
out of curiosity: this is basically `for i in range(len(F_0)):`, right?

gbergsten replied:
Yup! Since only the 'i' index is needed, `range(len(F_0))` is really all that's needed. I just have a nasty habit of using `enumerate` in all my loops...

matiscke reviewed (changes requested):
I tested the new function using the very useful example notebook, as well as with custom-built generator objects. For the applicable parameter space, this already seems to work! 
 
I see the following open points we should address before merging. 
- [x] Documentation: In the docs, we should update all references to the SAG13 to the new occurrence rates, adding reference(s) to the paper and specifics/limitations as needed. The changes to the docs could, but don't have to, be in their own commit.
- [x] Mstar range: I think we should implement a solution for the spectral types outside of the new function. Perhaps falling back to SAG13 in the case of `M_st_min < 0.56`, or covering only the stars below this mass with a different occurrence rate law. To be discussed.
- [x] Related: The error when `M_st_min < 0.56` should be caught and a custom error message raised so the user knows what the problem is. May not be needed if the above task is completed.
~~- [ ] Same for the radius range. Bioverse should produce giant planets, too, for which I understand we'll need a different source. Smaller planets could be extrapolated.~~
- [x] somehow the stellar masses generated are capped at ~0.7 Msun even though I set a much larger `M_st_max`. Is this a bug? 
![image](https://user-images.githubusercontent.com/19270173/194191988-74133fcd-e7a5-473c-a0a8-72e9154990b1.png)
- [x] let's discuss how we can make the new function work both when `transit_mode` is True and when it is False.

matiscke commented:
As discussed in meeting with @gbergsten and @matiscke, the 0.7Msun-issue is a bug in the (deprecated) `create_stars_Gaia` function and can be ignored for now.

gbergsten committed:
adding M-dwarf occurrence

Performed some extra fits to explore the occurrence of small planets around stars below 0.56 Msun.

gbergsten commented:
Added an extra mass bin for stars with `[0.010,0.556] Msun`, with an occurrence rate model fit to the planets orbiting these lowest-mass Kepler stars and informed by the next-heaviest mass bin. With this, `create_planets_bergsten` should no longer break when `M_st_min < 0.56 Msun`.

matiscke committed:
provide a transit mode for bergsten et al. occurrence rates

matiscke committed:
update SAG13 references to new occurrence rates
